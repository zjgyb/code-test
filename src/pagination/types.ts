/*
 * @Author: gyb
 * @Date: 2021-10-18 08:26:00
 * @Description: 分页组件
 */

import type { IxPublicPropTypes } from '../table/types';

// 分页组件的props
export const paginationProps = {
	current: {
		type: Number
	},
	total: {
		type: Number,
		default: 0
	},
	pageSize: {
		type: Number,
		default: 10
	}
};

// 页码组件
export const itemProps = {
	value: {
		type: Number,
		default: 1
	},
	activeIndex: {
		type: Number,
		default: 1
	}
}

export type PaginationPublicProps = Required<IxPublicPropTypes<typeof paginationProps>>
export type ItemPublicProps = Required<IxPublicPropTypes<typeof itemProps>>