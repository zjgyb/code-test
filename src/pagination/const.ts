/*
 * @Author: gyb
 * @Date: 2021-11-04
 * @Description: 分页组件常量地址
 */

export const middleCount = 5; // 中间页码个数
export const count = 2; // 第一页和最后一页固定死，一共两个
export const allCount = middleCount + count;