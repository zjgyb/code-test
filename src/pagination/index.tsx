/*
 * @Author: gyb
 * @Date: 2021-10-18
 * @Description: 分页组件导出
 */

import SfPagination from './pagination';

export { SfPagination };