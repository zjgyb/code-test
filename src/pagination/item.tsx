/*
 * @Author: gyb
 * @Date: 2021-10-18
 * @Description: 分页组件页码
 */

import type { ItemPublicProps } from './types';

import { computed, defineComponent } from '@vue/composition-api';
import { itemProps } from './types';

export default defineComponent({
  name: 'SfPaginationItem',
  props: itemProps,
  setup(props: ItemPublicProps, { emit }) {
    const classes = useClasses(props);
    const onClick = () => {
      emit('change', props.value);
    };

    return () => {
      return (
        <li class={classes.value} title={props.value?.toString()} onClick={onClick}>{props.value}</li>
      )
    }
  },
})

function useClasses(props: ItemPublicProps) {
  return computed(() => ({
    'sf-pagination-item': true,
    'sf-pagination-item--active': props.value === props.activeIndex
  }));
}