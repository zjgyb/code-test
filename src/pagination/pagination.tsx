/*
 * @Author: gyb
 * @Date: 2021-10-18
 * @Description: 分页组件
 */

import type { PaginationPublicProps } from './types';
import type { Ref, SetupContext } from '@vue/composition-api';

import { computed, defineComponent } from '@vue/composition-api';
import { paginationProps } from './types';
import PaginationItem from './item';

export default defineComponent({
	name: 'SfPagination',
	props: paginationProps,
	setup(props: PaginationPublicProps, context) {
		const endPage = computed(() => {
			return Math.ceil(props.total / props.pageSize);
		});
		const { prevCls, nextCls } = useClasses(props, endPage);
		const children = getRangeItems(props, endPage, context);
		const { onPageChange } = getEvents(props, endPage, context);

		return {
			endPage,
			prevCls,
			nextCls,
			children,
			onPageChange
		}
	},
	render() {
		return <ul class="sf-pagination">
			{/* 未国际化 */}
			<li class={this.prevCls} title="上一页" onClick={() => this.onPageChange(-1)}>
				&lt;
			</li>
			{
				this.children
			}
			<li class={this.nextCls} title="下一页" onClick={() => this.onPageChange()}>
				&gt;
			</li>
		</ul>
	}
})

function useClasses(props: PaginationPublicProps, endPage: Ref<number>) {
	const prevCls = computed(() => {
		return {
			'sf-pagination-item': true,
			'sf-pagination-item--disabled': props.current === 1
		}
	});
	const nextCls = computed(() => {
		return {
			'sf-pagination-item': true,
			'sf-pagination-item--disabled': props.current === endPage.value
		}
	})
	return {
		prevCls,
		nextCls
	}
}

function getItemsNumber(start: number, end: number) {
	let nums = [];
	for (let i = start; i < end; i++) {
		nums.push(i);
	}
	return nums;
}

function getRangeItems(props: PaginationPublicProps, endPage: Ref<number>, context: SetupContext) {
	return computed(() => {
		const middleCount = 5; // 中间页码个数
		const count = 2; // 第一页和最后一页固定死，一共两个
		const allCount = middleCount + count;
		const isOverMinCount = endPage.value > middleCount + count;
		const { emitPageNum, onPageChange } = getEvents(props, endPage, context);
		let itemPages: number[] = [];

		if (endPage.value <= allCount) {
			itemPages = getItemsNumber(count, endPage.value);
		} else {
			let prev = Math.max(count, props.current - count);
			prev = Math.min(prev, endPage.value - middleCount + 1);
			let end = Math.min(prev + middleCount, endPage.value);
			itemPages = getItemsNumber(prev, end);
		}

		// 判断是否需要显示往前5页与往后5页的按钮
		const hasPrev5 = isOverMinCount && (props.current >= middleCount);
		const haxNext5 = isOverMinCount && (props.current <= endPage.value - middleCount + 1)

		return [
			<PaginationItem active-index={props.current} onChange={emitPageNum} />,
			hasPrev5 && <li class="sf-pagination-item sf-pagination-item_prev" title={`向前${middleCount}页`} onClick={() => onPageChange(-middleCount)}></li>,
			...itemPages.map(i => {
				return <PaginationItem title={i} active-index={props.current} value={i} onChange={emitPageNum} />
			}),
			haxNext5 && <li class="sf-pagination-item sf-pagination-item_next" title={`向后${middleCount}页`} onClick={() => onPageChange(middleCount)}></li>,
			endPage.value > 1 && <PaginationItem value={endPage.value} active-index={props.current} onChange={emitPageNum} />
		]
	})
}

function getEvents(props: PaginationPublicProps, endPage: Ref<number>, context: SetupContext) {

	// 判断是否等于边界值
	const pageMap = new Map([
		[1, endPage.value],
		[-1, 1]
	]);

	const emitPageNum = (pageNum: number) => {
		context.emit('change', pageNum);
		context.emit('update:current', pageNum);
	}

	const onPageChange = (value = 1) => {
		if (props.current === pageMap.get(value)) {
			return;
		}
		let currentPage = props.current + value;
		currentPage = Math.min(Math.max(currentPage, 1), endPage.value);
		emitPageNum(currentPage);
	}

	return {
		emitPageNum,
		onPageChange
	}
}