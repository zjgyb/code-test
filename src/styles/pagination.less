/*
 * @Author: gyb
 * @Date: 2021-10-18
 * @Description: 分页器样式
 */

.sf-pagination {
  display: flex;

  &-item {
    min-width: 32px;
    height: 32px;
    line-height: 30px;
    border: 1px solid #f5f5f5;
    border-radius: 2px;
    text-align: center;
    cursor: pointer;

    &:not(:last-child) {
      margin-right: 8px;
    }

    &:hover,
    &--active {
      color: var(--active-color);
      border-color: var(--active-color);
    }

    &--disabled {

      &,
      &:hover {
        color: #eee;
        border-color: #d9d9d9;
        cursor: not-allowed;
      }

      user-select: none;
    }

    &_prev {
      &::after {
        content: "···";
      }

      &:hover {
        &::after {
          content: "<<";
        }
      }
    }

    &_next {
      &::after {
        content: "···";
      }

      &:hover {
        &::after {
          content: ">>";
        }
      }
    }
  }
}
