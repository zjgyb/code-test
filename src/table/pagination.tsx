/*
 * @Author: gyb
 * @Date: 2021-10-18
 * @Description: 表格分页相关代码
 */

import type { TablePaginationOptions } from './types';

import { SfPagination } from '../pagination';

export function renderPagination(options: TablePaginationOptions) {
    const { pagination, props, onChangePager } = options;
    
    if (props.pagination === false) {
        return null;
    }
    
    return <SfPagination class="sf-table-pagination" current={pagination.current} total={pagination.total} onChange={onChangePager} />
}