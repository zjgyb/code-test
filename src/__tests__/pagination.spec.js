import { mount } from '@vue/test-utils'
import { SfPagination } from '../pagination'

describe('Pagination', () => {
    const PaginationMount = options => mount(SfPagination, options)

    test('render', () => {
        const wrapper = PaginationMount()
        expect(wrapper.html()).toMatchSnapshot()
        expect(() => {
            wrapper.vm.$forceUpdate()
            wrapper.vm.$destroy()
        }).not.toThrow()
    })

    test('click', async() => {
        const wrapper = PaginationMount({
            propsData: {
                current: 1,
                total: 200
            }
        });
        const pagers = wrapper.findAll('.sf-pagination-item');
        pagers.at(2).trigger('click');
        const emits = wrapper.emitted()['change'];
        expect(emits).toBeTruthy();
        wrapper.setProps({ current: emits[0][0] });
        await wrapper.vm.$nextTick();
        expect(wrapper.find('.sf-pagination-item--active').text()).toBe('2');
    });

    test('hasPrev5', async() => {
        const wrapper = PaginationMount({
            propsData: {
                current: 5,
                total: 200
            }
        });
        const prev = wrapper.find('.sf-pagination-item_prev');
        expect(prev.exists()).toBeTruthy();
        prev.trigger('click');
        const emits = wrapper.emitted()['update:current'];
        expect(emits[0]).toEqual([1]);
        wrapper.setProps({ current: emits[0][0] });
        await wrapper.vm.$nextTick();
        expect(wrapper.find('.sf-pagination-item--active').text()).toBe('1');
        expect(wrapper.find('.sf-pagination-item_prev').exists()).toBeFalsy();
    })

    test('hasNext5', async() => {
        const wrapper = PaginationMount({
            propsData: {
                current: 1,
                total: 100
            }
        });
        const prev = wrapper.find('.sf-pagination-item_next');
        expect(prev.exists()).toBeTruthy();
        prev.trigger('click');
        const emits = wrapper.emitted()['update:current'];
        expect(emits[0]).toEqual([6]);
        wrapper.setProps({ current: emits[0][0] });
        await wrapper.vm.$nextTick();
        expect(wrapper.find('.sf-pagination-item--active').text()).toBe('6');
    })

    test('prev click', async() => {
        const wrapper = PaginationMount({
            propsData: {
                current: 1,
                total: 100
            }
        });
        let prevBtn = wrapper.find('.sf-pagination-item--disabled');
        expect(prevBtn.exists()).toBeTruthy()
        prevBtn.trigger('click');
        const emits = wrapper.emitted()['update:current'];
        expect(emits).toBeFalsy();
    });

    test('next click', async() => {
        const wrapper = PaginationMount({
            propsData: {
                current: 10,
                total: 100
            }
        });
        let nextBtns = wrapper.find('.sf-pagination-item--disabled');
        expect(nextBtns.exists()).toBeTruthy()
        nextBtns.trigger('click');
        const emits = wrapper.emitted()['update:current'];
        expect(emits).toBeFalsy();
    })
})