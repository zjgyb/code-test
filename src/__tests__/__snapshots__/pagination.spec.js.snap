// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination render 1`] = `
"<ul class=\\"sf-pagination\\">
  <li title=\\"上一页\\" class=\\"sf-pagination-item\\">&lt;</li>
  <li title=\\"1\\" class=\\"sf-pagination-item sf-pagination-item--active\\">1</li>
  <li title=\\"下一页\\" class=\\"sf-pagination-item\\">&gt;</li>
</ul>"
`;
